File: reference/lca.py
from pprint import pprint

class Node:
    def __init__(self, key):
        self.key = key
        self.children = []
        self.ancestor = None
        self.rank = 0
        self.color = 0
        self.parent = None

def printnodes():
    pprint(vars(p_u))
    pprint(vars(p_v))
    #pprint(vars(three))
    #pprint(vars(four))
    #pprint(vars(five))

def makeset(node):
    node.parent = node
    node.rank = 1

def union(nodex, nodey):
    xroot = find(nodex)
    yroot = find(nodey)
    if xroot.rank > yroot.rank:
        yroot.parent = xroot
    elif xroot.rank < yroot.rank:
        xroot.parent = yroot
    elif xroot.rank == yroot.rank:
        yroot.parent = xroot
        xroot.rank += 1

def find(node):
    if node.parent != node:
        node.parent = find(node.parent)
    return node.parent

def lca(u, found = False):
    #print "lca with %d" %(u.key)
    #printnodes()
    makeset(u)
    u.ancestor = u
    for w in u.children:
        res = lca(w, False)
        if res["found"]:
            return res
        #print "union u: %d, w: %d" %(u.key, w.key)
        union(u,w)
        #print "u.parent: %d, w.parent: %d, u.rank: %d, w.rank: %d" %(u.parent.key, w.parent.key, u.rank, w.rank)
        find(u).ancestor = u
        #print "find(u): %d, u_a: %d, w_a: %d, find(w).a: %d" %(find(u).key, u.ancestor.key, w.ancestor.key, find(w).ancestor.key)
    u.color = 1
    #print "p_u: %d, p_v: %d, u: %d, p_v.color: %d, p_u.color: %d, p_v_a: %d" %(p_u.key, p_v.key, u.key, p_v.color, p_u.color, (p_v.ancestor and p_v.ancestor.key) or 0)
    if (p_v.key == u.key) and (p_v.color == 1 and p_u.color == 1):
        #printnodes()
        #print find(p_u).key
        #print find(p_u).ancestor.key
        #print "\n"
        return {"a": find(p_u).ancestor, "found": True}
    elif (p_u.key == u.key) and (p_v.color == 1 and p_u.color == 1):
        #printnodes()
        #print find(p_v).key
        #print find(p_v).ancestor.key
        #print "hgf \n"
        return { "a": find(p_v).ancestor, "found": True}
    else:
        return {"found": False}

root = Node(1)
two = Node(2)
three = Node(3)
four = Node(4)
five = Node(5)
six = Node(6)
seven = Node(7)
eight = Node(8)

root.children.append(two)
root.children.append(three)
two.children.append(four)
two.children.append(five)
three.children.append(six)
three.children.append(seven)
three.children.append(eight)

p_u = two
p_v = eight

a = lca(root)
print a["a"].key

File: src/ccm/lib/lca/lca.go
package lca

type node struct {
	name string
	children []*node
	ancestor *node
	rank int
	color int
	parent *node
}

type Data struct {
	Name string
	Children []*Data
}

var root, target_node_1, target_node_2 *node

func makeset(n *node) {
	n.parent =  n
	n.rank = 1
}

func union(nodex, nodey *node) {
	xroot := find(nodex)
	yroot := find(nodey)
	if xroot.rank > yroot.rank {
		yroot.parent = xroot
	} else if xroot.rank < yroot.rank {
		xroot.parent = yroot
	} else if xroot.rank == yroot.rank {
		yroot.parent = xroot
		xroot.rank += 1
	}
}

func find(node *node) *node {
	if node.parent != node {
		node.parent = find(node.parent)
	}
	return node.parent
}

func lca(node *node) (*node, bool) {
	makeset(node)
	node.ancestor = node
	for _, child := range node.children {
		return_value, found := lca(child)
		if found == true {
			return return_value, found
		}
		union(node, child)
		this_node := find(node)
		this_node.ancestor = node
	}
	node.color = 1
	if node.name == target_node_2.name && target_node_2.color == 1 && target_node_1.color == 1 {
		return_val := find(target_node_1).ancestor
		return return_val, true
	} else if node.name == target_node_1.name && target_node_2.color == 1 && target_node_1.color == 1 {
		return_val := find(target_node_2).ancestor
		return return_val, true
	} else {
		return nil, false
	}
}

func find_node_by_name(name string, parent *node) (*node, bool) {
	if parent.name == name {
		return parent, true
	}
	for _, child := range parent.children {
		node, found := find_node_by_name(name, child)
		if found == true {
			return node, true
		}
	}
	return nil, false
}

func build_tree(data *Data) *node {
	node := &node{name: data.Name}
	for _,child := range data.Children {
		child_node := build_tree(child)
		node.children = append(node.children, child_node)
	}
	return node
}

func Search(name1, name2 string) string {
	target_node_1, _ = find_node_by_name(name1, root)
	target_node_2, _ = find_node_by_name(name2, root)
	ancestor, _ := lca(root)
	return ancestor.name
}

func Init(data *Data) {
	root = build_tree(data)
}
